File: android/sdk/src/main/java/com/viant/forgeandroid/runtime/ReportRuntimeStructure.kt
package com.viant.forgeandroid.runtime

import kotlinx.serialization.json.*

object ReportRuntimeStructure {
    fun composites(blocks: List<DashboardReportRuntimeBlockSummary>): List<DashboardReportRuntimeBlockSummary> {
        val index = blocks.associateBy { it.id }
        val parents = mutableMapOf<String, String>()
        val children = mutableMapOf<String, MutableList<String>>()
        for (block in blocks.filter { it.kind == "compositeBlock" }) {
            for (value in (block.content["childBlockIds"] as? JsonArray).orEmpty()) {
                val child = (value as? JsonPrimitive)?.takeIf { it.isString }?.content?.trim().orEmpty()
                if (child.isEmpty() || child !in index || child in parents) continue
                var ancestor: String? = block.id
                while (ancestor != null && ancestor != child) ancestor = parents[ancestor]
                if (ancestor == child) continue
                parents[child] = block.id
                children.getOrPut(block.id) { mutableListOf() }.add(child)
            }
        }
        fun attach(block: DashboardReportRuntimeBlockSummary): DashboardReportRuntimeBlockSummary = block.copy(
            compositeParentId = parents[block.id], children = children[block.id].orEmpty().mapNotNull(index::get).map(::attach)
        )
        return blocks.map(::attach)
    }
}

File: ios/Sources/ForgeIOSRuntime/ReportRuntimeStructure.swift
import Foundation

public enum ReportRuntimeStructure {
    public static func composites(_ blocks: [DashboardReportRuntimeBlockSummary]) -> [DashboardReportRuntimeBlockSummary] {
        let index = Dictionary(blocks.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })
        var parents: [String: String] = [:]
        var children: [String: [String]] = [:]
        for block in blocks where block.kind == "compositeBlock" {
            for value in block.content["childBlockIds"]?.arrayValue ?? [] {
                guard let child = value.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines), !child.isEmpty, index[child] != nil, parents[child] == nil else { continue }
                var ancestor: String? = block.id
                while let current = ancestor, current != child { ancestor = parents[current] }
                guard ancestor != child else { continue }
                parents[child] = block.id
                children[block.id, default: []].append(child)
            }
        }
        func attach(_ block: DashboardReportRuntimeBlockSummary) -> DashboardReportRuntimeBlockSummary {
            var result = block
            result.compositeParentID = parents[block.id]
            result.children = (children[block.id] ?? []).compactMap { index[$0] }.map(attach)
            return result
        }
        return blocks.map(attach)
    }
}
